File: cli/src/languages/cpp.js
import { LanguageAdapter } from './base.js';

/**
 * C / C++ language adapter.
 *
 * Walks a tree-sitter AST produced by either the C or C++ grammar and extracts
 * functions, #include imports, exports (non-static symbols), classes/structs,
 * and type declarations (struct, class, enum, typedef, namespace).
 *
 * A single adapter class handles both C and C++ because the C++ grammar is a
 * superset of C, and the extraction logic is largely identical.
 */
export class CppAdapter extends LanguageAdapter {
  constructor() {
    super('cpp');
  }

  // ---------------------------------------------------------------------------
  // Functions
  // ---------------------------------------------------------------------------

  /**
   * Extract function definitions.
   *
   * In C/C++ the AST node is `function_definition`.  Its `declarator` field
   * is a `function_declarator` whose own `declarator` gives the function name
   * (possibly qualified with `::` in C++).
   */
  extractFunctions(tree, sourceCode) {
    const functions = [];
    this._walkNodes(tree.rootNode, (node) => {
      if (node.type === 'function_definition') {
        const fn = this._parseFunctionDef(node, sourceCode);
        if (fn) functions.push(fn);
      }
    });
    return functions;
  }

  /**
   * Parse a function_definition node into a structured record.
   *
   * The signature is reconstructed from the return type, declarator text
   * (which includes the name and parameter list), and any trailing qualifiers.
   */
  _parseFunctionDef(node, sourceCode) {
    // Find the function_declarator – may be nested inside pointer_declarator
    // or reference_declarator.
    const funcDeclarator = this._findDescendantOfType(node, 'function_declarator');
    if (!funcDeclarator) return null;

    const nameNode = funcDeclarator.childForFieldName('declarator');
    if (!nameNode) return null;

    const name = nameNode.text;

    // Build signature from everything before the body
    const body = node.childForFieldName('body');
    let signature;
    if (body) {
      const sigEnd = body.startIndex;
      signature = sourceCode.slice(node.startIndex, sigEnd).trim();
    } else {
      signature = sourceCode.slice(node.startIndex, node.endIndex).trim();
    }

    return {
      name,
      signature,
      startLine: node.startPosition.row + 1,
      endLine: node.endPosition.row + 1,
    };
  }

  // ---------------------------------------------------------------------------
  // Imports (#include directives)
  // ---------------------------------------------------------------------------

  /**
   * Extract `#include` directives.
   *
   * - `#include <header>` → system_lib_string → isExternal: true
   * - `#include "header"` → string_literal    → isExternal: false
   */
  extractImports(tree, _sourceCode) {
    const imports = [];
    this._walkNodes(tree.rootNode, (node) => {
      if (node.type !== 'preproc_include') return;

      const pathNode = this._findChildOfType(node, 'system_lib_string')
        || this._findChildOfType(node, 'string_literal');
      if (!pathNode) return;

      const isExternal = pathNode.type === 'system_lib_string';
      // Strip < > or " "
      const raw = pathNode.text.replace(/^[<"]|[>"]$/g, '');

      imports.push({
        source: raw,
        symbols: [],
        isExternal,
      });
    });
    return imports;
  }

  // ---------------------------------------------------------------------------
  // Exports (non-static function and type names)
  // ---------------------------------------------------------------------------

  /**
   * In C/C++ there is no `export` keyword (ignoring C++20 modules).
   * By convention, all non-static functions and all type names at file /
   * namespace scope are considered "exported".
   */
  extractExports(tree, _sourceCode) {
    const exports = [];

    this._walkNodes(tree.rootNode, (node) => {
      // Non-static function definitions
      if (node.type === 'function_definition') {
        if (this._hasStorageClassStatic(node)) return;
        const funcDeclarator = this._findDescendantOfType(node, 'function_declarator');
        if (!funcDeclarator) return;
        const nameNode = funcDeclarator.childForFieldName('declarator');
        if (nameNode) {
          // Use the bare identifier (strip qualified names like Engine::start → start)
          const name = this._bareIdentifier(nameNode.text);
          exports.push(name);
        }
      }

      // Struct / class / enum type names
      if (
        node.type === 'struct_specifier' ||
        node.type === 'class_specifier' ||
        node.type === 'enum_specifier'
      ) {
        const nameNode = node.childForFieldName('name');
        if (nameNode) exports.push(nameNode.text);
      }

      // Typedefs
      if (node.type === 'type_definition') {
        const declarator = this._findDescendantOfType(node, 'type_identifier');
        if (declarator) exports.push(declarator.text);
      }
    });

    // Deduplicate (e.g. a struct may appear as both struct_specifier and in a typedef)
    return [...new Set(exports)];
  }

  // ---------------------------------------------------------------------------
  // Classes (C++ classes and C structs)
  // ---------------------------------------------------------------------------

  /**
   * Map C++ `class_specifier` and C/C++ `struct_specifier` nodes to the
   * "classes" output.  Only named specifiers with a body are included (forward
   * declarations are skipped).
   */
  extractClasses(tree, _sourceCode) {
    const classes = [];
    this._walkNodes(tree.rootNode, (node) => {
      if (node.type !== 'class_specifier' && node.type !== 'struct_specifier') return;

      const nameNode = node.childForFieldName('name');
      if (!nameNode) return;

      const body = node.childForFieldName('body');
      if (!body) return; // skip forward declarations

      classes.push({
        name: nameNode.text,
        startLine: node.startPosition.row + 1,
        endLine: node.endPosition.row + 1,
      });
    });
    return classes;
  }

  // ---------------------------------------------------------------------------
  // Types (struct, class, enum, typedef, namespace)
  // ---------------------------------------------------------------------------

  extractTypes(tree, _sourceCode) {
    const types = [];
    this._walkNodes(tree.rootNode, (node) => {
      // struct
      if (node.type === 'struct_specifier') {
        const nameNode = node.childForFieldName('name');
        const body = node.childForFieldName('body');
        if (nameNode && body) {
          types.push({
            name: nameNode.text,
            kind: 'struct',
            startLine: node.startPosition.row + 1,
            endLine: node.endPosition.row + 1,
          });
        }
      }

      // class (C++ only)
      if (node.type === 'class_specifier') {
        const nameNode = node.childForFieldName('name');
        const body = node.childForFieldName('body');
        if (nameNode && body) {
          types.push({
            name: nameNode.text,
            kind: 'class',
            startLine: node.startPosition.row + 1,
            endLine: node.endPosition.row + 1,
          });
        }
      }

      // enum (including C++ enum class)
      if (node.type === 'enum_specifier') {
        const nameNode = node.childForFieldName('name');
        if (nameNode) {
          types.push({
            name: nameNode.text,
            kind: 'enum',
            startLine: node.startPosition.row + 1,
            endLine: node.endPosition.row + 1,
          });
        }
      }

      // typedef
      if (node.type === 'type_definition') {
        const declarator = this._findDescendantOfType(node, 'type_identifier');
        if (declarator) {
          types.push({
            name: declarator.text,
            kind: 'typedef',
            startLine: node.startPosition.row + 1,
            endLine: node.endPosition.row + 1,
          });
        }
      }

      // namespace (C++ only)
      if (node.type === 'namespace_definition') {
        const nameNode = node.childForFieldName('name');
        if (nameNode) {
          types.push({
            name: nameNode.text,
            kind: 'namespace',
            startLine: node.startPosition.row + 1,
            endLine: node.endPosition.row + 1,
          });
        }
      }
    });
    return types;
  }

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  /** Find the first descendant (BFS) with the given type. */
  _findDescendantOfType(node, type) {
    const queue = [];
    for (let i = 0; i < node.childCount; i++) {
      queue.push(node.child(i));
    }
    while (queue.length > 0) {
      const current = queue.shift();
      if (current.type === type) return current;
      for (let i = 0; i < current.childCount; i++) {
        queue.push(current.child(i));
      }
    }
    return null;
  }

  /**
   * Check whether a function_definition has the `static` storage class.
   *
   * In tree-sitter-c/cpp, the storage class specifier appears as a direct
   * child of the function_definition with type `storage_class_specifier`
   * whose text is "static".
   */
  _hasStorageClassStatic(funcDefNode) {
    for (let i = 0; i < funcDefNode.childCount; i++) {
      const child = funcDefNode.child(i);
      if (child.type === 'storage_class_specifier' && child.text === 'static') {
        return true;
      }
    }
    return false;
  }

  /**
   * Extract the bare identifier from a possibly qualified name.
   * e.g. "Engine::start" → "start", "initialize" → "initialize"
   */
  _bareIdentifier(text) {
    const idx = text.lastIndexOf('::');
    return idx >= 0 ? text.slice(idx + 2) : text;
  }
}

File: rust-cli/src/languages/cpp.rs
use tree_sitter::{Language, Tree};
use super::{
    ClassInfo, ExportInfo, FunctionInfo, ImportInfo, LanguageAdapter,
    find_descendant_of_type, node_text, walk_nodes,
};
use super::c_lang::{extract_c_includes, extract_c_functions, extract_c_exports, extract_c_classes};

pub struct CppAdapter;

impl CppAdapter {
    pub fn new() -> Self {
        Self
    }
}

impl LanguageAdapter for CppAdapter {
    fn language(&self) -> Language {
        tree_sitter_cpp::LANGUAGE.into()
    }

    fn extract_functions(&self, tree: &Tree, source: &[u8]) -> Vec<FunctionInfo> {
        // C++ 函数提取与 C 相同，但名称可能含 :: 限定符
        extract_c_functions(tree, source)
    }

    fn extract_imports(&self, tree: &Tree, source: &[u8]) -> Vec<ImportInfo> {
        extract_c_includes(tree, source)
    }

    fn extract_exports(&self, tree: &Tree, source: &[u8]) -> Vec<ExportInfo> {
        extract_c_exports(tree, source)
    }

    fn extract_classes(&self, tree: &Tree, source: &[u8]) -> Vec<ClassInfo> {
        let mut classes = extract_c_classes(tree, source);
        // 额外处理 C++ 类方法
        walk_nodes(tree.root_node(), &mut |node| {
            if node.kind() != "class_specifier" {
                return;
            }
            if node.child_by_field_name("body").is_none() {
                return;
            }
            if let Some(name_node) = node.child_by_field_name("name") {
                let class_name = node_text(name_node, source).to_string();
                if let Some(ci) = classes.iter_mut().find(|c| c.name == class_name && c.kind == "class") {
                    ci.methods = extract_cpp_methods(node, source);
                }
            }
        });
        // 添加 enum（与 Node.js 一致）
        walk_nodes(tree.root_node(), &mut |node| {
            if node.kind() != "enum_specifier" {
                return;
            }
            if node.child_by_field_name("body").is_none() {
                return;
            }
            if let Some(n) = node.child_by_field_name("name") {
                classes.push(ClassInfo {
                    name: node_text(n, source).to_string(),
                    start_line: node.start_position().row + 1,
                    end_line: node.end_position().row + 1,
                    methods: Vec::new(),
                    kind: "enum".into(),
                });
            }
        });
        // 添加 namespace（与 Node.js 一致）
        walk_nodes(tree.root_node(), &mut |node| {
            if node.kind() != "namespace_definition" {
                return;
            }
            if let Some(n) = node.child_by_field_name("name") {
                classes.push(ClassInfo {
                    name: node_text(n, source).to_string(),
                    start_line: node.start_position().row + 1,
                    end_line: node.end_position().row + 1,
                    methods: Vec::new(),
                    kind: "namespace".into(),
                });
            }
        });
        classes
    }
}

fn extract_cpp_methods(class_node: tree_sitter::Node, source: &[u8]) -> Vec<String> {
    let mut methods = Vec::new();
    walk_nodes(class_node, &mut |node| {
        if node.kind() == "function_definition" {
            if let Some(func_decl) = find_descendant_of_type(node, "function_declarator") {
                if let Some(name_node) = func_decl.child_by_field_name("declarator") {
                    methods.push(node_text(name_node, source).to_string());
                }
            }
        }
    });
    methods
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(source: &str) -> tree_sitter::Tree {
        let adapter = CppAdapter::new();
        let mut parser = tree_sitter::Parser::new();
        parser.set_language(&adapter.language()).unwrap();
        parser.parse(source, None).unwrap()
    }

    #[test]
    fn test_cpp_extract_functions() {
        let src = r#"
#include <string>

class Engine {
public:
    void start() {}
    void stop() {}
};

int main() {
    return 0;
}
"#;
        let tree = parse(src);
        let adapter = CppAdapter::new();
        let fns = adapter.extract_functions(&tree, src.as_bytes());
        assert!(fns.iter().any(|f| f.name == "main"));
    }

    #[test]
    fn test_cpp_extract_includes() {
        let src = "#include <vector>\n#include \"engine.h\"\n";
        let tree = parse(src);
        let adapter = CppAdapter::new();
        let imports = adapter.extract_imports(&tree, src.as_bytes());
        assert!(imports.iter().any(|i| i.source == "vector" && i.is_default));
        assert!(imports.iter().any(|i| i.source == "engine.h" && !i.is_default));
    }

    #[test]
    fn test_cpp_extract_classes() {
        let src = r#"
class Animal {
public:
    void speak() {}
};

struct Point {
    int x, y;
};
"#;
        let tree = parse(src);
        let adapter = CppAdapter::new();
        let classes = adapter.extract_classes(&tree, src.as_bytes());
        assert!(classes.iter().any(|c| c.name == "Animal" && c.kind == "class"));
        assert!(classes.iter().any(|c| c.name == "Point" && c.kind == "struct"));
    }

    #[test]
    fn test_cpp_namespace_exports() {
        let src = r#"
namespace MyLib {
    void helper() {}
}
"#;
        let tree = parse(src);
        let adapter = CppAdapter::new();
        let exports = adapter.extract_exports(&tree, src.as_bytes());
        // Node.js 行为：namespace 本身不出现在 exports 中
        assert!(!exports.iter().any(|e| e.name == "MyLib" && e.kind == "namespace"),
            "namespace should not appear in exports");
    }
}
